### DISCUSSION TITLE
64 x 64 Matrix using virtual display library

### DISCUSSION THREAD

**Discussion by Gavo4** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#674] [Q&A]
Hi There,

First of all the libraries are great!, good job!,

Secondly I ma having an issue with the ESP32 Trinity when I am connected 2 x 64x64 Panels 2 x 2; if I connect then 1 x 4 in series the standard examples work great!, the minute I try and have 2 on the top and 2 on the bottom things start getting weird, its as if the libraries cater only for 64 x 32 panels. Can anyone point my in the right direction?

**Comment by Gavo4** [comment]
![Clipboard](https://github.com/user-attachments/assets/d3c279c8-38ee-48e9-9319-3bdae6cfbddb)
This is what I am seeing

**Reply by davemaster** [reply]
> ![Clipboard](https://private-user-images.githubusercontent.com/25132911/366058209-d3c279c8-38ee-48e9-9319-3bdae6cfbddb.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMDY2MTIsIm5iZiI6MTcyNjAwNjMxMiwicGF0aCI6Ii8yNTEzMjkxMS8zNjYwNTgyMDktZDNjMjc5YzgtMzhlZS00OGU5LTkzMTktM2JkYWU2Y2ZiZGRiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDIyMTE1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwNjhiMzFjNzM4MjJjMDhjYjQ5YjM5MWE1Y2E3MWMyZWQ4NTVlYmYzNzQ2NzZkZDc1ODIyOTM2NTRhMjE5NjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CFcDUxlDbsBdDKwSBbFaMUnNzFvlEYbNvq-MIvAhgFg) This is what I am seeing

but, CODE please...

**Comment by board707** [comment]
Is your 64x64 panel 8 or 16 scans with complex pattern?

**Comment by Gavo4** [comment]
`/******************************************************************************
    Configuration for a virtual display made up of four 64x64 panels in a 2x2 grid
*****************************************************************************/

// 1) Include key virtual display library
#include <ESP32-VirtualMatrixPanel-I2S-DMA.h>
#include <FastLED.h>


#define R1 25
#define G1 26
#define BL1 27
#define R2 14
#define G2 12
#define BL2 13
#define CH_A 23
#define CH_B 19
#define CH_C 5
#define CH_D 17
#define CH_E -1 // assign to any available pin if using two panels or 64x64 panels with 1/32 scan
#define CLK 16
#define LAT 4
#define OE 15
#define PANEL_RES_X 64  // Width of each panel
#define PANEL_RES_Y 64  // Height of each panel
#define NUM_ROWS 2  // Number of rows of chained individual panels.
#define NUM_COLS 2  // Number of individual panels per row.
#define PANEL_CHAIN NUM_ROWS * NUM_COLS  // Total number of panels chained

#define PIN_E 18


/* Configure the serpentine chaining approach. Options are:
    CHAIN_TOP_LEFT_DOWN
    CHAIN_TOP_RIGHT_DOWN
    CHAIN_BOTTOM_LEFT_UP
    CHAIN_BOTTOM_RIGHT_UP

    In this example, we'll use 'CHAIN_BOTTOM_LEFT_UP' which starts from the bottom left.
*/
#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_LEFT_DOWN
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_RIGHT_DOWN
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_LEFT_UP
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_RIGHT_UP

// 3) Create the runtime objects to use

// Placeholder for the matrix object.
MatrixPanel_I2S_DMA *dma_display = nullptr;

// Placeholder for the virtual display object.
VirtualMatrixPanel  *virtualDisp = nullptr;

/******************************************************************************
 * Setup!
 ******************************************************************************/
void setup() {
  
  delay(2000);
  Serial.begin(115200);
  Serial.println(""); Serial.println(""); Serial.println("");
  Serial.println("*****************************************************");
  Serial.println(" HELLO !");
  Serial.println("*****************************************************");

  // Create the physical DMA panel class and virtual (chained) display class.   
  HUB75_I2S_CFG mxconfig(
                PANEL_RES_X,   // Module width
                PANEL_RES_Y,   // Module height
                PANEL_CHAIN    // Chain length
  );

  // Set the panel type if using a specific chipset (like FM6126A)
  // mxconfig.driver = HUB75_I2S_CFG::FM6126A;     

  // Check for the need for Virtual Display class
  if (NUM_ROWS <= 1) {
    Serial.println(F("There is no reason to use the VirtualDisplay class for a single horizontal chain and row!"));
  }

  // Create the DMA display object
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);

  // Set brightness to about 75%
  dma_display->setBrightness8(192);  // Range is 0-255

  // Allocate memory and start DMA display
  if( not dma_display->begin() )
      Serial.println("****** !KABOOM! I2S memory allocation failed ***********");

  // Create the Virtual Display object
  virtualDisp = new VirtualMatrixPanel((*dma_display), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, VIRTUAL_MATRIX_CHAIN_TYPE);

  // Clear the screen
  virtualDisp->fillScreen(virtualDisp->color444(0, 0, 0));
  
  // Draw initial test pattern to verify panel connections
  virtualDisp->drawDisplayTest();

  Serial.println("Chain of 4x 64x64 panels for this example:");
  Serial.println("+---------+---------+");
  Serial.println("|    4    |    3    |");
  Serial.println("|         |         |");
  Serial.println("+---------+---------+");
  Serial.println("|    2    |    1     |");
  Serial.println("|         | (ESP32)  |");
  Serial.println("+---------+---------+");

// Set up the font and text properties
virtualDisp->setFont(&FreeSansBold12pt7b);
virtualDisp->setTextColor(virtualDisp->color565(0, 0, 255));
virtualDisp->setTextSize(2);  // Adjust text size for visibility on 64x64 matrix

// Center the text on the 64x64 display
const char* text = "ABCD";
int textWidth = 8 * 16; // Assume an average character width of 12 pixels for size 2 (adjust as needed)
int textHeight = 20;    // Assume text height of 24 pixels for size 2 (adjust as needed)

int xPos = (virtualDisp->width() - textWidth) / 2;  // Calculate X position for centering
int yPos = (virtualDisp->height() - textHeight) / 2; // Calculate Y position for centering

virtualDisp->setCursor(xPos, yPos);
virtualDisp->print(text);

// Draw a corrected red rectangle around the entire 64x64 display
virtualDisp->drawRect(0, 0, virtualDisp->width(), virtualDisp->height(), virtualDisp->color565(255, 0, 0));

// Move the green boundary box down by 16 rows
int boxWidth = virtualDisp->width();
int boxHeight = 0; // Height of the green boundary box (adjust if needed)

virtualDisp->drawRect(0, 0, boxWidth, boxHeight, virtualDisp->color565(0, 255, 0)); // Starting at row 16 to move it down by 16 rows

// Draw a white diagonal line from the top left to the bottom right
virtualDisp->drawLine(0, 0, virtualDisp->width() - 1, virtualDisp->height() - 1, virtualDisp->color565(255, 255, 255));

// Optional: Draw a diagonal line from the bottom left to the top right for a cross effect
virtualDisp->drawLine(0, virtualDisp->height() - 1, virtualDisp->width() - 1, 0, virtualDisp->color565(255, 255, 255));

// Redraw test pattern to check panel connectivity
virtualDisp->drawDisplayTest();

}

void loop() {
  // No loop action needed for a static display
}
Here is the code

**Comment by davemaster** [comment]
Greetings,

this line, tells a lot:

#define CH_E -1 // assign to any available pin if using two panels or 64x64 panels with 1/32 scan


You are using a 64x64 panel, in 32S, You need the E-Pin assigned to a GPIO pin, NECESSARY, obliged.... 

This code below, is my config for 64x64 32S panel

```
#include <ESP32-HUB75-MatrixPanel-I2S-DMA.h>

/*--------------------- MATRIX GPIO CONFIG  -------------------------*/
/*--------------------- HUB75 MATRIX GPIO CONFIG  ---------------------*/
#define R1  2
#define G1  23
#define BL1  4
#define R2  16
#define G2  27
#define BL2  17

#define CH_A   5
#define CH_B   18
#define CH_C   19
#define CH_D   21
#define CH_E   33
#define LAT 26
#define OE  25

#define CLK 22


/*--------------------- MATRIX PANEL CONFIG -------------------------*/
#define PANEL_RES_X 64      // Number of pixels wide of each INDIVIDUAL panel module. 
#define PANEL_RES_Y 64     // Number of pixels tall of each INDIVIDUAL panel module.
#define PANEL_CHAIN 1      // Total number of panels chained one to another

MatrixPanel_I2S_DMA *dma_display = nullptr;

// Module configuration
HUB75_I2S_CFG::i2s_pins _pins={R1, G1, BL1, R2, G2, BL2, CH_A, CH_B, CH_C, CH_D, CH_E, LAT, OE, CLK};
HUB75_I2S_CFG mxconfig(
                PANEL_RES_X,    // module width
                PANEL_RES_Y,    // module height
                PANEL_CHAIN,    // chain length
                _pins           // pin mapping      
);

//mxconfig.gpio.E = 33; // Assign a pin if you have a 64x64 panel
//mxconfig.clkphase = false; // Change this if you have issues with ghosting.
//mxconfig.driver = HUB75_I2S_CFG::FM6126A; // Change this according to your panel.



#include <FastLED.h>
#define MATRIX_WIDTH 64
#define MATRIX_HEIGHT 64
```

etc

**Comment by board707** [comment]
@davemaster 
You are absolutely right, for matrices with 32 scans pin E is mandatory.

It is not clear how @Gavo4 managed to make everything work when connecting panels in one line:
> if I connect then 1 x 4 in series the standard examples work great!,

**Reply by Gavo4** [reply]
Connecting in one line or connection 4 in series is no issue, the ESP32 Trinity uses the E line on GPIO18 on the ESP32, here is the code I used to connect 4 in a line which works great!, the issue is connecting two at the bottom and 2 at the top

**Comment by Gavo4** [comment]
`#define R1 25
#define G1 26
#define BL1 27
#define R2 14
#define G2 12
#define BL2 13
#define CH_A 23
#define CH_B 19
#define CH_C 5
#define CH_D 17
#define CH_E -1 // assign to any available pin if using two panels or 64x64 panels with 1/32 scan
#define CLK 16
#define LAT 4
#define OE 15


#include <ESP32-HUB75-MatrixPanel-I2S-DMA.h>
#include <FastLED.h>

// Configure for your panel(s) as appropriate!
#define PANEL_WIDTH 64
#define PANEL_HEIGHT 64  	// Panel height of 64 will required PIN_E to be defined.
#define PANELS_NUMBER 2	// Number of chained panels, if just a single panel, obviously set to 1
#define PIN_E 18

#define PANE_WIDTH PANEL_WIDTH * PANELS_NUMBER
#define PANE_HEIGHT PANEL_HEIGHT


// placeholder for the matrix object
MatrixPanel_I2S_DMA *dma_display = nullptr;


uint16_t time_counter = 0, cycles = 0, fps = 0;
unsigned long fps_timer;

CRGB currentColor;
CRGBPalette16 palettes[] = {HeatColors_p, LavaColors_p, RainbowColors_p, RainbowStripeColors_p, CloudColors_p};
CRGBPalette16 currentPalette = palettes[0];


CRGB ColorFromCurrentPalette(uint8_t index = 0, uint8_t brightness = 255, TBlendType blendType = LINEARBLEND) {
  return ColorFromPalette(currentPalette, index, brightness, blendType);
}

void setup() {
  
  Serial.begin(115200);
  
  Serial.println(F("*****************************************************"));
  Serial.println(F("*        ESP32-HUB75-MatrixPanel-I2S-DMA DEMO       *"));
  Serial.println(F("*****************************************************"));


  
  HUB75_I2S_CFG mxconfig;
  mxconfig.mx_height = PANEL_HEIGHT;      // we have 64 pix heigh panels
  mxconfig.chain_length = PANELS_NUMBER;  // we have 2 panels chained
  mxconfig.gpio.e = PIN_E;                // we MUST assign pin e to some free pin on a board to drive 64 pix height panels with 1/32 scan
  //mxconfig.driver = HUB75_I2S_CFG::FM6126A;     // in case that we use panels based on FM6126A chip, we can change that


  // OK, now we can create our matrix object
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);

  // let's adjust default brightness to about 75%
  dma_display->setBrightness8(255);    // range is 0-255, 0 - 0%, 255 - 100%

  // Allocate memory and start DMA display
  if( not dma_display->begin() )
      Serial.println("****** !KABOOM! I2S memory allocation failed ***********");
 
  // well, hope we are OK, let's draw some colors first :)
  Serial.println("Fill screen: RED");
  dma_display->fillScreenRGB888(255, 0, 0);
  delay(1000);

  Serial.println("Fill screen: GREEN");
  dma_display->fillScreenRGB888(0, 255, 0);
  delay(1000);

  Serial.println("Fill screen: BLUE");
  dma_display->fillScreenRGB888(0, 0, 255);
  delay(1000);

  Serial.println("Fill screen: Neutral White");
  dma_display->fillScreenRGB888(64, 64, 64);
  delay(1000);

  Serial.println("Fill screen: black");
  dma_display->fillScreenRGB888(0, 0, 0);
  delay(1000);


  // Set current FastLED palette
  currentPalette = RainbowColors_p;
  Serial.println("Starting plasma effect...");
  fps_timer = millis();
}

void loop() {
  
    for (int x = 0; x < PANE_WIDTH; x++) {
            for (int y = 0; y <  PANE_HEIGHT; y++) {
                int16_t v = 128;
                uint8_t wibble = sin8(time_counter);
                v += sin16(x * wibble * 3 + time_counter);
                v += cos16(y * (128 - wibble)  + time_counter);
                v += sin16(y * x * cos8(-time_counter) / 8);

                currentColor = ColorFromPalette(currentPalette, (v >> 8)); //, brightness, currentBlendType);
                dma_display->drawPixelRGB888(x, y, currentColor.r, currentColor.g, currentColor.b);
            }
    }

    ++time_counter;
    ++cycles;
    ++fps;

    if (cycles >= 1024) {
        time_counter = 0;
        cycles = 0;
        currentPalette = palettes[random(0,sizeof(palettes)/sizeof(palettes[0]))];
    }

    // print FPS rate every 5 seconds
    // Note: this is NOT a matrix refresh rate, it's the number of data frames being drawn to the DMA buffer per second
    if (fps_timer + 5000 < millis()){
      Serial.printf_P(PSTR("Effect fps: %d\n"), fps/5);
      fps_timer = millis();
      fps = 0;
    }
} // end loop`

**Comment by Gavo4** [comment]
I also tried this code with the no luck, this display looks like it is looking for a 64 x 32 matrix `/******************************************************************************
    Configuration for a virtual display made up of four 64x64 panels in a 2x2 grid
*****************************************************************************/

// 1) Include key virtual display library
#include <ESP32-VirtualMatrixPanel-I2S-DMA.h>
#include <FastLED.h>


#define R1 25
#define G1 26
#define BL1 27
#define R2 14
#define G2 12
#define BL2 13
#define CH_A 23
#define CH_B 19
#define CH_C 5
#define CH_D 17
#define CH_E 18 // assign to any available pin if using two panels or 64x64 panels with 1/32 scan
#define CLK 16
#define LAT 4
#define OE 15
#define PANEL_RES_X 64  // Width of each panel
#define PANEL_RES_Y 64  // Height of each panel
#define NUM_ROWS 2  // Number of rows of chained individual panels.
#define NUM_COLS 2  // Number of individual panels per row.
#define PANEL_CHAIN NUM_ROWS * NUM_COLS  // Total number of panels chained

#define PIN_E 18


/* Configure the serpentine chaining approach. Options are:
    CHAIN_TOP_LEFT_DOWN
    CHAIN_TOP_RIGHT_DOWN
    CHAIN_BOTTOM_LEFT_UP
    CHAIN_BOTTOM_RIGHT_UP

    In this example, we'll use 'CHAIN_BOTTOM_LEFT_UP' which starts from the bottom left.
*/
#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_LEFT_DOWN
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_RIGHT_DOWN
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_LEFT_UP
//#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_RIGHT_UP

// 3) Create the runtime objects to use

// Placeholder for the matrix object.
MatrixPanel_I2S_DMA *dma_display = nullptr;

// Placeholder for the virtual display object.
VirtualMatrixPanel  *virtualDisp = nullptr;

/******************************************************************************
 * Setup!
 ******************************************************************************/
void setup() {
  
  delay(2000);
  Serial.begin(115200);
  Serial.println(""); Serial.println(""); Serial.println("");
  Serial.println("*****************************************************");
  Serial.println(" HELLO !");
  Serial.println("*****************************************************");

  // Create the physical DMA panel class and virtual (chained) display class.   
  HUB75_I2S_CFG mxconfig(
                PANEL_RES_X,   // Module width
                PANEL_RES_Y,   // Module height
                PANEL_CHAIN    // Chain length
  );

  // Set the panel type if using a specific chipset (like FM6126A)
  // mxconfig.driver = HUB75_I2S_CFG::FM6126A;     

  // Check for the need for Virtual Display class
  if (NUM_ROWS <= 1) {
    Serial.println(F("There is no reason to use the VirtualDisplay class for a single horizontal chain and row!"));
  }

  // Create the DMA display object
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);

  // Set brightness to about 75%
  dma_display->setBrightness8(192);  // Range is 0-255

  // Allocate memory and start DMA display
  if( not dma_display->begin() )
      Serial.println("****** !KABOOM! I2S memory allocation failed ***********");

  // Create the Virtual Display object
  virtualDisp = new VirtualMatrixPanel((*dma_display), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, VIRTUAL_MATRIX_CHAIN_TYPE);

  // Clear the screen
  virtualDisp->fillScreen(virtualDisp->color444(0, 0, 0));
  
  // Draw initial test pattern to verify panel connections
  virtualDisp->drawDisplayTest();

// Set up the font and text properties
virtualDisp->setFont(&FreeSansBold12pt7b);
virtualDisp->setTextColor(virtualDisp->color565(0, 0, 255));
virtualDisp->setTextSize(2);  // Adjust text size for visibility on 64x64 matrix

// Center the text on the 64x64 display
const char* text = "ABCD";
int textWidth = 8 * 16; // Assume an average character width of 12 pixels for size 2 (adjust as needed)
int textHeight = 20;    // Assume text height of 24 pixels for size 2 (adjust as needed)

int xPos = (virtualDisp->width() - textWidth) / 2;  // Calculate X position for centering
int yPos = (virtualDisp->height() - textHeight) / 2; // Calculate Y position for centering

virtualDisp->setCursor(xPos, yPos);
virtualDisp->print(text);

// Draw a corrected red rectangle around the entire 64x64 display
virtualDisp->drawRect(0, 0, virtualDisp->width(), virtualDisp->height(), virtualDisp->color565(255, 0, 0));

// Move the green boundary box down by 16 rows
int boxWidth = virtualDisp->width();
int boxHeight = 0; // Height of the green boundary box (adjust if needed)

virtualDisp->drawRect(0, 0, boxWidth, boxHeight, virtualDisp->color565(0, 255, 0)); // Starting at row 16 to move it down by 16 rows

// Draw a white diagonal line from the top left to the bottom right
virtualDisp->drawLine(0, 0, virtualDisp->width() - 1, virtualDisp->height() - 1, virtualDisp->color565(255, 255, 255));

// Optional: Draw a diagonal line from the bottom left to the top right for a cross effect
virtualDisp->drawLine(0, virtualDisp->height() - 1, virtualDisp->width() - 1, 0, virtualDisp->color565(255, 255, 255));

// Redraw test pattern to check panel connectivity
virtualDisp->drawDisplayTest();

}

void loop() {
  // No loop action needed for a static display
}
`

**Reply by davemaster** [reply]
> I also tried this code with the no luck, this display looks like it is looking for a 64 x 32 matrix `/****************************************************************************** Configuration for a virtual display made up of four 64x64 panels in a 2x2 grid *****************************************************************************/
> 
> // 1) Include key virtual display library #include <ESP32-VirtualMatrixPanel-I2S-DMA.h> #include <FastLED.h>
> 
> #define R1 25 #define G1 26 #define BL1 27 #define R2 14 #define G2 12 #define BL2 13 #define CH_A 23 #define CH_B 19 #define CH_C 5 #define CH_D 17 #define CH_E 18 // assign to any available pin if using two panels or 64x64 panels with 1/32 scan #define CLK 16 #define LAT 4 #define OE 15 #define PANEL_RES_X 64 // Width of each panel #define PANEL_RES_Y 64 // Height of each panel #define NUM_ROWS 2 // Number of rows of chained individual panels. #define NUM_COLS 2 // Number of individual panels per row. #define PANEL_CHAIN NUM_ROWS * NUM_COLS // Total number of panels chained
> 
> #define PIN_E 18
> 
> /* Configure the serpentine chaining approach. Options are: CHAIN_TOP_LEFT_DOWN CHAIN_TOP_RIGHT_DOWN CHAIN_BOTTOM_LEFT_UP CHAIN_BOTTOM_RIGHT_UP
> 
> ```
> In this example, we'll use 'CHAIN_BOTTOM_LEFT_UP' which starts from the bottom left.
> ```
> 
> */ #define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_LEFT_DOWN //#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_RIGHT_DOWN //#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_LEFT_UP //#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_BOTTOM_RIGHT_UP
> 
> // 3) Create the runtime objects to use
> 
> // Placeholder for the matrix object. MatrixPanel_I2S_DMA *dma_display = nullptr;
> 
> // Placeholder for the virtual display object. VirtualMatrixPanel *virtualDisp = nullptr;
> 
> /******************************************************************************
> 
>     * Setup!
>       ******************************************************************************/
>       void setup() {
> 
> 
> delay(2000); Serial.begin(115200); Serial.println(""); Serial.println(""); Serial.println(""); Serial.println("_**"); Serial.println(" HELLO !"); Serial.println("**_");
> 
> // Create the physical DMA panel class and virtual (chained) display class. HUB75_I2S_CFG mxconfig( PANEL_RES_X, // Module width PANEL_RES_Y, // Module height PANEL_CHAIN // Chain length );
> 
> // Set the panel type if using a specific chipset (like FM6126A) // mxconfig.driver = HUB75_I2S_CFG::FM6126A;
> 
> // Check for the need for Virtual Display class if (NUM_ROWS <= 1) { Serial.println(F("There is no reason to use the VirtualDisplay class for a single horizontal chain and row!")); }
> 
> // Create the DMA display object dma_display = new MatrixPanel_I2S_DMA(mxconfig);
> 
> // Set brightness to about 75% dma_display->setBrightness8(192); // Range is 0-255
> 
> // Allocate memory and start DMA display if( not dma_display->begin() ) Serial.println("****** !KABOOM! I2S memory allocation failed ***********");
> 
> // Create the Virtual Display object virtualDisp = new VirtualMatrixPanel((*dma_display), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, VIRTUAL_MATRIX_CHAIN_TYPE);
> 
> // Clear the screen virtualDisp->fillScreen(virtualDisp->color444(0, 0, 0));
> 
> // Draw initial test pattern to verify panel connections virtualDisp->drawDisplayTest();
> 
> // Set up the font and text properties virtualDisp->setFont(&FreeSansBold12pt7b); virtualDisp->setTextColor(virtualDisp->color565(0, 0, 255)); virtualDisp->setTextSize(2); // Adjust text size for visibility on 64x64 matrix
> 
> // Center the text on the 64x64 display const char* text = "ABCD"; int textWidth = 8 * 16; // Assume an average character width of 12 pixels for size 2 (adjust as needed) int textHeight = 20; // Assume text height of 24 pixels for size 2 (adjust as needed)
> 
> int xPos = (virtualDisp->width() - textWidth) / 2; // Calculate X position for centering int yPos = (virtualDisp->height() - textHeight) / 2; // Calculate Y position for centering
> 
> virtualDisp->setCursor(xPos, yPos); virtualDisp->print(text);
> 
> // Draw a corrected red rectangle around the entire 64x64 display virtualDisp->drawRect(0, 0, virtualDisp->width(), virtualDisp->height(), virtualDisp->color565(255, 0, 0));
> 
> // Move the green boundary box down by 16 rows int boxWidth = virtualDisp->width(); int boxHeight = 0; // Height of the green boundary box (adjust if needed)
> 
> virtualDisp->drawRect(0, 0, boxWidth, boxHeight, virtualDisp->color565(0, 255, 0)); // Starting at row 16 to move it down by 16 rows
> 
> // Draw a white diagonal line from the top left to the bottom right virtualDisp->drawLine(0, 0, virtualDisp->width() - 1, virtualDisp->height() - 1, virtualDisp->color565(255, 255, 255));
> 
> // Optional: Draw a diagonal line from the bottom left to the top right for a cross effect virtualDisp->drawLine(0, virtualDisp->height() - 1, virtualDisp->width() - 1, 0, virtualDisp->color565(255, 255, 255));
> 
> // Redraw test pattern to check panel connectivity virtualDisp->drawDisplayTest();
> 
> }
> 
> void loop() { // No loop action needed for a static display } `

is this, all your code??? Send me please this complete to test, also a HQ Picture or video of your connection. Thanks in advance

**Reply by Gavo4** [reply]
Hi Dave,

That's all I have, I agree it looks like it's looking for a 64 x 32, I was just trying to get the Matrix's to show something. I am using the Waveshare P3 RGB 64 x 64 Led Panels with the ESP32 Trinity which seems to work well if the panels are connecting in series but placing them in a 2 x 2 format does not seem to work

**Comment by davemaster** [comment]
yes, goal, see something on panel.. even a single led....

I will try this in a couple of hours in my own 64x64 P3 panel...

**Reply by Gavo4** [reply]
Thanks so much!, any help would be great, greatly appreciated!

**Comment by davemaster** [comment]
Well,

I tested your example (test001.ino) with some modifications, and works perfectly, bellow both: video and code.

Best regards... 

PD.  I SUGGEST, check the connections, each cable, your power supply, 5V 5A or more...


[the code - TEST_001.ino](https://github.com/davemaster/LEDBOARD-ESP32-HUB-75-TEST-CODE/blob/58c8fc4528baf6ccb9cca0102acd9c575916580f/Examples/TEST_001.ino)

[the video - TEST 001 ESP32 HUB75 DMA 64X64 32S](https://youtu.be/MAhkiFnunmE)

And please, a like and subscription will be great... 

Thanks in advance

**Reply by board707** [reply]
> I tested your example (test001.ino) with some modifications,

Hi @davemaster 
As I see, you tested this with a single panel in code settings:
```
#define PANELS_NUMBER 1 
```
Unfortunately, this devalues ​​your entire work, since the OP's problem is not in working with 64x64 panels in general, but specifically in outputting an image onto 4 such panels, connected as 2x2 set.

**Comment by Gavo4** [comment]
Thanks David,

This works for a single panel but the minute I try and add 2 x 64 x 64 in a matrix of 4 this does not work, I have tried everything possible to get the 2 x 64 top and bottom to work, I can get 4 x 64 to work in a row but not ontop of each other. All panel do light up on the 2 x 2 but the screen are not in Sync i can either get the 2 at the top to work together and the 2 at the bottom but no all 4 to work in sync

**Reply by board707** [reply]
Hi @Gavo4 
Could it be a memory issue?

I am going also to test this too with four 64x64 panels in setting. Unfortunately, I have little time now. I will try to do it tonight, if not then next week, if the problem is not solved by then.

**Reply by Gavo4** [reply]
I don't believe so I am using an ESP32 and also it works when 4 are connected in series, it just when I stack them

**Reply by davemaster** [reply]
> No lo creo, estoy usando un ESP32 y también funciona cuando 4 están conectados en serie, solo cuando los apilo.

Well, now, You must use the examples included in this library... wax on - wax off
![imagen](https://github.com/user-attachments/assets/ead547bc-7fd8-409b-9cd7-942eecf9c8f6)

**Comment by davemaster** [comment]
Search in examples, and is there, only there, you can find "the master"

<img src="https://github.com/user-attachments/assets/c7c8e6eb-0c42-451f-9d3b-09da874314c9" width="300">

A sample.... there, only there

![Imagen de WhatsApp 2024-09-12 a las 11 13 53_c0cdf2ca](https://github.com/user-attachments/assets/7bf48a80-8515-48c8-9eae-2f4a7948a439)

**Comment by board707** [comment]
@davemaster 
It is not clear, what your picture mean. Did you again tested a code for a single panel? What is the point of this?
If you trying to help, why don't you test a code with 4 panels?

@Gavo4 
> I don't believe so I am using an ESP32 

See the Readme - the maximum guaranteed screen size is 128x64, while you try to use 128x128.

**Reply by davemaster** [reply]
> @davemaster It is not clear, what your picture mean. Did you again tested a code for a single panel? What is the point of this? If you trying to help, why don't you test a code with 4 panels?
> 
> @Gavo4
> 
> > I don't believe so I am using an ESP32
> 
> See the Readme - the maximum guaranteed screen size is 128x64, while you try to use 128x128.

I mean, "... what we helped You, can find that can use VirtualPanel from Examples... they're right there.... Arduino IDE->Files->Examples->Customized Libraries->ESP32-HUB75---- BLAH BLAH

![imagen](https://github.com/user-attachments/assets/0f071637-d710-46f4-ae50-4b509c654385)

**Comment by Gavo4** [comment]
Hi All,

I solved it, here is the code, 

`#include <ESP32-VirtualMatrixPanel-I2S-DMA.h>
#include <Fonts/FreeSansBold12pt7b.h>  // Include the font

#define PANEL_RES_X 64
#define PANEL_RES_Y 64

#define NUM_ROWS 2
#define NUM_COLS 2
#define PANEL_CHAIN NUM_ROWS * NUM_COLS

#define R1 25
#define G1 26
#define BL1 27
#define R2 14
#define G2 12
#define BL2 13
#define CH_A 23
#define CH_B 19
#define CH_C 5
#define CH_D 17
#define CH_E 18 // Set to -1 if not used
#define CLK 16
#define LAT 4
#define OE 15

#define VIRTUAL_MATRIX_CHAIN_TYPE CHAIN_TOP_RIGHT_DOWN

MatrixPanel_I2S_DMA *dma_display = nullptr;
VirtualMatrixPanel *virtualDisp = nullptr;

void setup() {
    delay(2000);
    Serial.begin(115200);
    Serial.println("Initializing...");

    HUB75_I2S_CFG mxconfig(PANEL_RES_X, PANEL_RES_Y, PANEL_CHAIN);
    mxconfig.gpio.e = CH_E;

    if (NUM_ROWS <= 1) {
        Serial.println(F("No need for VirtualDisplay for a single chain and row!"));
    }

    dma_display = new MatrixPanel_I2S_DMA(mxconfig);
    if (!dma_display) {
        Serial.println("Failed to initialize DMA display object!");
        while (1);
    }

    dma_display->setBrightness8(192);

    if (!dma_display->begin()) {
        Serial.println("I2S memory allocation failed.");
        while (1);
    }

    virtualDisp = new VirtualMatrixPanel(*dma_display, NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, VIRTUAL_MATRIX_CHAIN_TYPE);
    if (!virtualDisp) {
        Serial.println("Failed to initialize VirtualMatrixPanel object!");
        while (1);
    }

    // Clear the screen
    virtualDisp->fillScreen(virtualDisp->color444(0, 0, 0));

    Serial.println("Chain of 4x 64x64 panels for this example:");
    Serial.println("+---------+---------+");
    Serial.println("|    2    |    1    |");
    Serial.println("|         | (ESP)   |");
    Serial.println("+---------+---------+");
    Serial.println("|    3    |    4    |");
    Serial.println("|         |         |");
    Serial.println("+---------+---------+");

    virtualDisp->setFont(&FreeSansBold12pt7b);
    virtualDisp->setTextColor(virtualDisp->color565(0, 0, 255));
    virtualDisp->setTextSize(3); 

    // Draw "A" in Panel 1 (Top Right)
    virtualDisp->setCursor(75, 55);
    virtualDisp->print("A");

    // Draw "B" in Panel 2 (Top Left)
    virtualDisp->setCursor(13, 55);
    virtualDisp->print("B");

    // Draw "C" in Panel 3 (Bottom Left)
    virtualDisp->setCursor(5, 120); 
    virtualDisp->print("C");

    // Draw "D" in Panel 4 (Bottom Right)
    virtualDisp->setCursor(64, 120); 
    virtualDisp->print("D");

    // Draw the top border (one row down, one column to the right)
    virtualDisp->drawRect(2, 0, 126, 64, virtualDisp->color565(255, 0, 0)); // Corrected to start from (1, 0) with width 126 and height 63

    // Draw the bottom border with adjustments for the rotated panels
    virtualDisp->drawRect(0, 64, 127, 64, virtualDisp->color565(255, 0, 0)); // Adjusted to align correctly with the rotated panels

    // Correct the diagonal line
    virtualDisp->drawLine(1, 0, 128, 128, virtualDisp->color565(255, 255, 255)); // Ensure the line runs from top-left to bottom-right
}

void loop() {
    // Your main loop
}
`

**Comment by davemaster** [comment]
Gavo4, share a video of your accomplishing 

Thanks in advance
